src/coreclr/jit/morph.cpp

```diff
@@ -14282,6 +14282,22 @@ GenTree* Compiler::fgMorphMultiOp(GenTreeMultiOp* multiOp)
     }
 #endif // defined(FEATURE_HW_INTRINSICS) && defined(TARGET_XARCH)
 
+#ifdef FEATURE_HW_INTRINSICS
+    if (multiOp->OperIsHWIntrinsic())
+    {
+        GenTreeHWIntrinsic* hw = multiOp->AsHWIntrinsic();
+
+        // See if this is foldable
+        GenTree* optTree = gtFoldHWIntrinsic(hw);
+
+        // If we optimized, morph the result
+        if (optTree != hw)
+        {
+            return fgMorphTree(optTree);
+        }
```

Comment:
We're in post-order at this point, and it is an error, generally speaking, to call `fgMorphTree` in post-order. We don't need (and don't want) to remorph the tree that we have just morphed, in most cases.

Reply:
This pattern is already done for `gtFoldExprCall` - so my assumption is that it's ok?

Reply:
> so my assumption is that it's ok?

Well, it is done by `fgMorphCall` in pre-order.

To expand on this a bit. Pre-order transforms that "expand" trees into some new sequences _must_ call `fgMorphTree` on the new tree (otherwise some nodes end up being not morphed, which is not legal).

Post-order transforms are generally expected to maintain the "morphness" of the tree (not introduce operations that need morphing for correctness). "Re-morphing" trees in post-order is a cumbersome thing, even if legal, because morph takes special precautions (in the form of the `GTF_DEBUG_NODE_MORPHED` flag) that are supposed to catch "double morphing". "To see it for yourself", you can just try `return fgMorphTree(tree)` somewhere in post-order - it will assert.

That's also why post-order that creates new trees such as this method needs to mark it with `GTF_DEBUG_NODE_MORPHED`.

Currently, "it is fine" because a new leaf node is returned, but as soon as something already morphed is going to show up, it will start to assert.

Reply:
(Might be a dumb question) How do I know if something is going to be in pre-order or post-order morph? I think this optimization could technically be fine in either, but I wonder which one is more "correct" in the sense that it can be expanded upon to handle other types of hw intrinsic nodes.

Reply:
Optimizations should be done in post-order unless there are "special" reasons for them not to be, as post-order knows more about the tree's operands (e. g. pre-order will miss trees that will later be folded into constants).

Reply:
After having learned the terms for "pre-order" and "post-order" and of what this means in the context of morph, this optimization happens in post-order as the children of these nodes are morphed by this point.